src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

```diff
@@ -0,0 +1,529 @@
+                JavaScript.Runtime.FreeObject(_onOpen);
+            }
+            if (_onMessage != null)
+            {
+                _innerWebSocket?.SetObjectProperty("onmessage", "");
+                JavaScript.Runtime.FreeObject(_onMessage);
+            }
+            _innerWebSocket?.Dispose();
+        }
+
+        // This method is registered by the CancellationTokenSource cts in the connect method
+        // and called by Dispose or Abort so that any open websocket connection can be closed.
+        private void AbortRequest()
+        {
+            if (State == WebSocketState.Open)
+            {
+                var closeResult = CloseAsyncCore(WebSocketCloseStatus.NormalClosure, "Connection was aborted", CancellationToken.None);
+            }
+        }
+
+        /// <summary>
+        /// Send data on <see cref="System.Net.WebSockets.ClientWebSocket"/> as an asynchronous operation.
+        /// </summary>
+        /// <returns>The async.</returns>
+        /// <param name="buffer">Buffer.</param>
+        /// <param name="messageType">Message type.</param>
+        /// <param name="endOfMessage">If set to <c>true</c> end of message.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        public override async Task SendAsync(ArraySegment<byte> buffer, WebSocketMessageType messageType, bool endOfMessage, CancellationToken cancellationToken)
+        {
+            ThrowIfNotConnected();
+
+            if (messageType != WebSocketMessageType.Binary &&
+                messageType != WebSocketMessageType.Text)
+            {
+                throw new ArgumentException($"Invalid message type: '{messageType}' specified in method 'SendAsync'.  Valid types are 'Binary' and 'Text'",
+                    nameof(messageType));
+            }
+
+            if (!endOfMessage && buffer.Array == null)
+                return;
+
+            if (!endOfMessage)
+            {
+                if (buffer.Array == null)
+                    return;
+                _writeBuffer = _writeBuffer ?? new MemoryStream();
+                _writeBuffer.Write(buffer.Array, buffer.Offset, buffer.Count);
+                return;
+            }
+            else if (_writeBuffer != null && buffer.Array != null)
+            {
+                _writeBuffer.Write(buffer.Array, buffer.Offset, buffer.Count);
+                if (!_writeBuffer.TryGetBuffer(out buffer))
+                    throw new WebSocketException(WebSocketError.NativeError);
+            }
+
+            var tcsSend = new TaskCompletionSource<bool>();
+            // Wrap the cancellationToken in a using so that it can be disposed of whether
+            // we successfully send or not.
+            // Otherwise any timeout/cancellation would apply to the full session.
+            var writtenBuffer = _writeBuffer;
+            _writeBuffer = null;
+
+            using (cancellationToken.Register(() => tcsSend.TrySetCanceled()))
```

Comment:
I don't understand the purpose of this. This is going to cause the tcs to be canceled, but it's not actually preventing any of the sending work from happening, and it's only be awaited after all the work has been done.

Reply:
Addressed in recent commits